src/topmodel.c

```diff
@@ -508,9 +508,189 @@ if(yes_print_output==TRUE)
 return;
 }
 
+
+/*********************************
+ * Define functions to be used in init() and in Set_value in bmi_topmodel.c
+ * to enable calibratable parameters to be updated
+ * **********************************/
+
+
+// FUNCTION TO CONVERT DISTANCE/AREA FORM TO TIME DELAY HISTOGRAM ORDINATES
+// outputs used in subsequent functions include tch
+extern void convert_dist_to_histords(double *dist_from_outlet, int num_channels,
+					double chv, double rv, double dt, double tch[11])
+{
+    // declare local variables
+    double chvdt, rvdt;
+    int j;
+    
+    /*  Convert parameters to m/time step DT
+    *  with exception of XK0 which must stay in m/h
+    *                    Q0 is already in m/time step
+                     T0 is input as Ln(To) */
+
+    chvdt = chv*dt;
+    rvdt = rv*dt;
+
+    tch[1]=dist_from_outlet[1]/chvdt;
+    for(j=2;j<=num_channels;j++)
+     {
+      tch[j]=tch[1]+(dist_from_outlet[j]-dist_from_outlet[1])/rvdt;
+     }
+
+    return;
+
+}
+
+// FUNCTION TO CALCULATE THE TIME_DELAY_HISTOGRAM
+// updates values associated with model structure
+// outputs used in TOPMOD() include num_time_delay_histo_ords, time_delay_histogram,
+// num_delay, and Q
+extern void calc_time_delay_histogram(int max_time_delay_ordinates, int num_channels, double area,
+				double tch[11], int *num_time_delay_histo_ords, int *num_delay,
+				double **time_delay_histogram, double *cum_dist_area_with_dist) 
+
+
+{
+
+    // declare local variables
+    double time, a1, sumar, a2;
+    int j, ir;
+
+    if((*time_delay_histogram)==NULL)
+      {
+      d_alloc(time_delay_histogram,max_time_delay_ordinates);
+      }
+
+
+    (*num_time_delay_histo_ords)=(int)tch[num_channels];
+    
+    
+    if((double)(*num_time_delay_histo_ords)<tch[num_channels]) 
+       {
+       (*num_time_delay_histo_ords)++;
+       }
+    
+    (*num_delay)=(int)tch[1];
+    (*num_time_delay_histo_ords)-=(*num_delay);
+    
+
+
+    for(ir=1;ir<=(*num_time_delay_histo_ords);ir++)
+      {
+      time=(double)(*num_delay)+(double)ir;
```

Comment:
again, this cast is suspicious, why does time need to be a double?  both num_delay and ir are ints, so should time simply be an int as well?

Reply:
The only thing I can think of is that `time` was expected to be large, and a double type was used to avoid overflowing an integer with large enough `time` values?  For signed (positive/negative) ints, an 8 bit int can hold numbers up to 127, a 16 bit int 32,767, 32-bit int 2,147,483,647, and a 64-bit int 9,223,372,036,854,775,807.

It is possible that original code was developed on something with only 16 bit integers and double was used to increase the largest possible time representation.

If we really want to ensure we can hold large representations for time, we can change the type of time to `int64_t` and use
```C
#include <stdint.h>
```
to have a portable, compliant 64 bit int to put that into...

We can also use unsigned types if we know the values are not ever going to be negative.

The other point I'll make here is that even if `time` does need to be a floating point, you don't have to cast the int, you can safely add ints and store them in a double, the compiler will handle that just fine.